virtualizarr/zarr.py

```diff
@@ -25,90 +25,50 @@
 
 
 class Codec(BaseModel):
-    compressor: str | None = None
+    """
+    ZarrayV2 codec definition.
+    """
+
+    compressor: str | dict[str, Any] | None = None
+    """
+    If it's a string, it's the compressor ID.
+    If it's a dict, it's the full compressor configuration.
+    """
     filters: list[dict] | None = None
 
     def __repr__(self) -> str:
         return f"Codec(compressor={self.compressor}, filters={self.filters})"
```

Comment:
Should this class be moved upstream?

Reply:
I'm not sure, because ZarrV3 now has a notion of [codecs](https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html#id12) which is different from VirtualiZarr's notion of codecs. A VirtualiZarr codec is really the combination of `filters` and `compressor` from V2 metadata, which are easily transformable to V3 codecs.

Maybe a `zarr.ArrayV2Metadata` can have a property that projects the `compressor` and `filters` into V3 `Codec`s, @d-v-b thoughts on that? In that case, `Codec` could even be a property of the ABC `ArrayMetadata`.

Reply:
I actually think that `codecs: ListWithInternalStructure[Codec]` (i.e., what zarr v3 metadata uses) is not the best interface. See a writeup on that point of view [here](https://github.com/zarr-developers/zarr-specs/issues/298). 

But if we ignore that concern for a moment, I'm curious about @TomNicholas's question: what could we change in zarr-python so that you could just import a functionally equivalent class from there (or import a dataclass and wrap with pydantic)? I have a selfish interest in this question since I'm likely going to follow in your footsteps when I implement better codec support over in [`pydantic-zarr`](https://github.com/janelia-cellmap/pydantic-zarr/blob/a1f7a978a5c68b8c3425be7d83202986ef828d12/src/pydantic_zarr/v3.py#L117).

Reply:
Thanks for forcing me to read the Zarr code more closely! I found VirtualiZarr could probably just use [create_codec_pipeline](https://github.com/zarr-developers/zarr-python/blob/b8baa6868c8fa95e6b2948c9fd9c725331ca23ec/src/zarr/array.py#L79C1-L87C29 ) to turn V2 or V3 metadata into a `BatchedCodecPipeline` and then VirtualiZarr can just deal with `CodecPipeline` as the abstraction over V2 and V3 . Additionally, VirtualiZarr could change its `Codec` to use [`V2Filters`](https://github.com/zarr-developers/zarr-python/blob/v3/src/zarr/codecs/_v2.py#L65) and `V2Compressor` from `Zarr`, or I might get rid of it entirely.

Reply:
what does Virtualizarr do with the codecs?

Reply:
All virtualizarr currently needs to do with the codecs is:
1) Check that any two arrays you ask it to concatenate have the same codecs (see https://github.com/zarr-developers/VirtualiZarr/issues/5)
2) Make sure the codecs are recorded in the kerchunk references when they get written out to disk.